### PR TITLE
Fixes #2699 Configurable View Timeout: SG responding gracefully when …

### LINKED ISSUE #2699: Configurable View Timeout: SG responding gracefully when the data is being rebuild should be configurable (enable/disable)
### Sync Gateway version
1.4.1

 ### Expected Behavior:
SG responding gracefully during data rebuild should be configurable to enable or disable this feature. When disabled, it should fallback to existing behavior. 
 
 ### Actual Behavior
 Requests to syncgateway nodes are getting hung/not responding/timeout during couchbase views rebuilding

### PR DESCRIPTION
**Pull request by tleyden**
Fixes https://github.com/couchbase/sync_gateway/issues/2699

…the data is being rebuild should be configurable (enable/disable)

### CODE REVIEW

**Comment by tleyden** [comment]
@adamcfraser ready for review

File: rest/config.go [hunk 1]
```diff
@@ -112,22 +112,23 @@ func (c ClusterConfig) CBGTEnabled() bool {
...
+	AllowEmptyPassword bool                           `json:"allow_empty_password,omitempty"`  // Allow empty passwords?  Defaults to false
+	CacheConfig        *CacheConfig                   `json:"cache,omitempty"`                 // Cache settings
+	ChannelIndex       *ChannelIndexConfig            `json:"channel_index,omitempty"`         // Channel index settings
+	RevCacheSize       *uint32                        `json:"rev_cache_size,omitempty"`        // Maximum number of revisions to store in the revision cache
+	StartOffline       bool                           `json:"offline,omitempty"`               // start the DB in the offline state, defaults to false
+	Unsupported        *UnsupportedConfig             `json:"unsupported,omitempty"`           // Config for unsupported features
+	OIDCConfig         *auth.OIDCOptions              `json:"oidc,omitempty"`                  // Config properties for OpenID Connect authentication
+	ViewQueryTimeoutMs *uint32                        `json:"view_query_timeout_ms,omitempty"` // The view query timeout in milliseconds
```
Comment (adamcfraser) [inline rest/config.go]: Do you think it's required to set the view query timeout in milliseconds in the config?  It seems more user-friendly to use seconds for this kind of timeout.  Was this specifically for QE's benefit?

**Review by adamcfraser — COMMENTED** [review]
Generally looks good - minor question about s vs ms in the config value.

File: rest/config.go (hunk 1 above)
Comment (tleyden) [reply rest/config.go]: Good catch, will change to seconds

**Comment by tleyden** [comment]
@adamcfraser changed units from ms -> secs

File: base/bucket.go [hunk 2]
```diff
@@ -54,8 +54,25 @@ type BucketSpec struct {
 	Server, PoolName, BucketName, FeedType string
 	Auth                                   AuthHandler
 	CouchbaseDriver                        CouchbaseDriver
-	MaxNumRetries                          int // max number of retries before giving up
-	InitialRetrySleepTimeMS                int // the initial time to sleep in between retry attempts (in millisecond), which will double each retry
+	MaxNumRetries                          int     // max number of retries before giving up
+	InitialRetrySleepTimeMS                int     // the initial time to sleep in between retry attempts (in millisecond), which will double each retry
+	ViewQueryTimeoutSecs                   *uint32 // the view query timeout in milliseconds (default: 75000ms)
```
Comment (adamcfraser) [inline base/bucket.go]: Comment still needs to be updated to seconds.

File: base/bucket.go [hunk 3]
```diff
@@ -54,8 +54,25 @@ type BucketSpec struct {
...
+func (b BucketSpec) GetViewQueryTimeout() time.Duration {
+
+	// If the user doesn't specify any timeout, default to 75s
+	if b.ViewQueryTimeoutSecs == nil {
+		return time.Second * 75
+	}
+
+	// If the user specifies 0, then translate that to "No timeout"
+	if *b.ViewQueryTimeoutSecs == 0 {
+		return time.Hour * 24 * 7 * 365 * 10 // 10 years
```
Comment (adamcfraser) [inline base/bucket.go]: I think this is actually 70 years, but don't think it really matters in this context.

**Pull request merged**